vowpalwabbit/slim/include/vw_slim_predict.h

```diff
@@ -0,0 +1,447 @@
+        break;
+      }
+      case vw_predict_exploration::bag:
+      {
+        std::vector<uint32_t> top_actions(num_actions);
+
+        // apply stride shifts
+        std::vector<std::unique_ptr<stride_shift_guard>> stride_shift_guards;
+				stride_shift_guards.push_back(std::unique_ptr<stride_shift_guard>(new stride_shift_guard(shared, _stride_shift)));
+        example_predict* actions_end = actions + num_actions;
+        for (example_predict* action = actions; action != actions_end; ++action)
+					stride_shift_guards.push_back(std::unique_ptr<stride_shift_guard>(new stride_shift_guard(*action, _stride_shift)));
+
+				for (size_t i = 0; i < _bag_size; i++)
+				{
+					std::vector<std::unique_ptr<feature_offset_guard>> feature_offset_guards;
+					for (example_predict* action = actions; action != actions_end; ++action)
+						feature_offset_guards.push_back(std::unique_ptr<feature_offset_guard>(new feature_offset_guard(*action, i)));
+
+					RETURN_ON_FAIL(predict(shared, actions, num_actions, scores));
+
+					auto top_action_iterator = std::min_element(std::begin(scores), std::end(scores));
+					uint32_t top_action = (uint32_t)(top_action_iterator - std::begin(scores));
+
+					top_actions[top_action]++;
+				}
+
+				// generate exploration distribution
+				RETURN_EXPLORATION_ON_FAIL(exploration::generate_bag(std::begin(top_actions), std::end(top_actions), std::begin(pdf), std::end(pdf)));
+
+        if (_minimum_epsilon > 0)
+					RETURN_EXPLORATION_ON_FAIL(exploration::enforce_minimum_probability(_minimum_epsilon, true, std::begin(pdf), std::end(pdf)));
+
+        break;
+      }
+      default:
+        return E_VW_PREDICT_ERR_NOT_A_CB_MODEL;
+    }
+
+		RETURN_EXPLORATION_ON_FAIL(sort_by_scores(std::begin(pdf), std::end(pdf), std::begin(scores), std::end(scores), std::begin(ranking), std::end(ranking)));
+
+    // Sample from the pdf
+    uint32_t chosen_action_idx;
+		RETURN_EXPLORATION_ON_FAIL(exploration::sample_after_normalizing(event_id, std::begin(pdf), std::end(pdf), chosen_action_idx));
+
+    // Swap top element with chosen one (unless chosen is the top)
+    if (chosen_action_idx != 0)
+    {
+      std::iter_swap(std::begin(ranking), std::begin(ranking) + chosen_action_idx);
+      std::iter_swap(std::begin(pdf), std::begin(pdf) + chosen_action_idx);
+    }
+
+    return S_VW_PREDICT_OK;
+  }
+
+		template<typename PdfIt, typename InputScoreIt, typename OutputIt>
+		static int sort_by_scores(	PdfIt pdf_first, PdfIt pdf_last,
+							InputScoreIt scores_first, InputScoreIt scores_last,
+      OutputIt ranking_begin, OutputIt ranking_last)
+  {
+    size_t pdf_size = pdf_last - pdf_first;
+    size_t ranking_size = ranking_last - ranking_begin;
+
+    if (pdf_size != ranking_size)
+      return E_EXPLORATION_PDF_RANKING_SIZE_MISMATCH;
+
+    // Intialize ranking with actions 0,1,2,3 ...
+    std::iota(ranking_begin, ranking_last, 0);
+
+    // Sort actions based on scores
+    std::sort(ranking_begin, ranking_last,
```

Comment:
```suggestion
    std::stable_sort(ranking_begin, ranking_last,
```

Reply:
Perhaps a new PR into this repo with this change and a couple of unit tests that verify the right behavior?  Is this something you mind doing as part of the investigation on empty model investigation?

---
In reply to: [304039901](https://github.com/VowpalWabbit/vowpal_wabbit/pull/1800#discussion_r304039901) [](ancestors = 304039901)

Reply:
Yes that’s much better. Will do